### DISCUSSION TITLE
Support persistent layouts in Vue 3's setup sugar

### DISCUSSION THREAD

**Discussion by innocenzi** [inertiajs/inertia#651] [Ideas]
When using the `setup` sugar ([RFC](https://github.com/vuejs/rfcs/blob/script-setup-2/active-rfcs/0000-script-setup.md)), we can't use persistent layouts, because there is no way to set a property (`layout` in our case) on the instance.

The only workaround I have right now is to abuse `defineProps`, which is used by the compiler: 
```html
<script setup lang="ts">
import layout from '@/views/layouts/main.vue'
import { defineProps } from 'vue'

defineProps({ layout })
</script>
``` 

With this, I have to update `resolveComponent` and I can use persistent layouts: 
```ts
createApp({
	render: () => h(app, {
		initialPage: JSON.parse(root.dataset.page!),
		resolveComponent: async(name: string) => {
			const { default: component } = (await import(`../views/pages/${name}.vue`))
			component.layout ??= component.props?.layout

			return component
		},
	}),
}).use(inertia).mount(root)
```

That's obviously an ugly workaround, and I'd like to be able to have a normal `resolveComponent` and having this kind of page component: 

```html
<script setup lang="ts">
import layout from '@/views/layouts/main.vue'
import { defineLayout } from '@inertiajs/inertia-vue3'

defineLayout(layout)
</script>
``` 

I can't see how to do that yet though. We can't even implement the ugly workaround by having a very specific property name, because `defineProps` is only available inside the `<script setup>` block, it can't be abstracted away, since the compiler specifically looks for it (ie. you don't actually need to `import { defineProps }  from 'vue'`).

**Comment by reinink** [comment]
Very interesting. Thanks for digging into this @innocenzi. We'll definitely want to solve this in a nice way, because `<script setup>` is awesome. 👍

**Comment by innocenzi** [comment]
This is probably fixed with https://github.com/vuejs/rfcs/blob/master/active-rfcs/0040-script-setup.md#exposing-components-public-interface now 👍 

---

Looks like this works: 

![](https://i.imgur.com/OrkSxAh.png)

But this doesn't:

![](https://i.imgur.com/uQokb6N.png)
![](https://i.imgur.com/2ufPqlr.png)

**EDIT** - The compiler warning is just my outdated `@vue/compiler-sfc`, but after an update it still does not work.

**Reply by innocenzi** [reply]
Okay, so I re-read the RFC and actually `defineExpose` is not made for that - it's for usage with template refs. So with Inertia we'll have to use a new `<script>` with a default export, or the `defineProps` hack.

Read: https://github.com/vuejs/rfcs/blob/master/active-rfcs/0040-script-setup.md#usage-alongside-normal-script

**Reply by innocenzi** [reply]
So, TL;DR: 

## Normal way

![](https://i.imgur.com/t1bMw8Z.png)

With that, no need to modify `resolveComponent`.

## With lazy-loading

![](https://i.imgur.com/HRBQIt0.png)
![](https://i.imgur.com/ue7Wuoz.png)

## Shorter syntax

![](https://i.imgur.com/INYJmtM.png)
![](https://i.imgur.com/ixdmiDE.png)

**Reply by reinink** [reply]
Awesome, thanks for figuring this out @innocenzi!

**Comment by xiCO2k** [comment]
Any update on having a `defineLayout` from the `@inertiajs/inertia-vue3`?

Thanks

**Reply by innocenzi** [reply]
As far as I know, this cannot happen without a PR to Vue.

**Reply by xiCO2k** [reply]
@innocenzi That's a bummer. I have the workaround you suggested, but does not look clean anymore.

**Reply by innocenzi** [reply]
Yep. If you are using Vite, you can easily preprocess your SFCs to have a cleaner solution though. https://twitter.com/enzoinnocenzi/status/1443171187190808578?t=RQF8dpZAllByFubIYN-nmQ&s=19

**Reply by xiCO2k** [reply]
I'm using laravel-mix, but since its a new project I may try vite. Thanks a lot

**Reply by innocenzi** [reply]
Another update, which Laravel Mix users can take advantage of, this time! https://twitter.com/enzoinnocenzi/status/1494177242074828803
